pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,488 @@
+	}
+
+	if err := addDataVolumeSnapshotCloneControllerWatches(mgr, dataVolumeCloneController); err != nil {
+		return nil, err
+	}
+
+	return dataVolumeCloneController, nil
+}
+
+func addDataVolumeSnapshotCloneControllerWatches(mgr manager.Manager, datavolumeController controller.Controller) error {
+	if err := addDataVolumeControllerCommonWatches(mgr, datavolumeController, dataVolumeSnapshotClone); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile loop for the clone data volumes
+func (r SnapshotCloneReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("DataVolume", req.NamespacedName)
+	return r.updateStatus(r.sync(log, req))
+}
+
+func (r SnapshotCloneReconciler) prepare(syncRes *dataVolumeSyncResult) error {
+	dv := syncRes.dvMutated
+	if err := r.populateSourceIfSourceRef(dv); err != nil {
+		return err
+	}
+	if dv.Status.Phase == cdiv1.Succeeded {
+		if err := r.cleanup(syncRes); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r SnapshotCloneReconciler) updateAnnotations(dataVolume *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) error {
+	if dataVolume.Spec.Source.Snapshot == nil {
+		return errors.Errorf("no source set for clone datavolume")
+	}
+	sourceNamespace := dataVolume.Spec.Source.Snapshot.Namespace
+	if sourceNamespace == "" {
+		sourceNamespace = dataVolume.Namespace
+	}
+	token, ok := dataVolume.Annotations[cc.AnnCloneToken]
+	if !ok {
+		return errors.Errorf("no clone token")
+	}
+	pvc.Annotations[cc.AnnCloneToken] = token
+	pvc.Annotations[cc.AnnCloneRequest] = sourceNamespace + "/" + dataVolume.Spec.Source.Snapshot.Name
+	return nil
+}
+
+func (r SnapshotCloneReconciler) sync(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	syncRes, syncErr := r.syncSnapshotClone(log, req)
+	if err := r.syncUpdate(log, &syncRes.dataVolumeSyncResult); err != nil {
+		syncErr = err
+	}
+	return syncRes, syncErr
+}
+
+func (r SnapshotCloneReconciler) syncSnapshotClone(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	res, syncErr := r.syncCommon(log, req, r.cleanup, r.prepare)
+	syncRes := dataVolumeCloneSyncResult{dataVolumeSyncResult: *res}
+	if syncErr != nil || syncRes.result != nil {
+		return syncRes, syncErr
+	}
+
+	pvc := syncRes.pvc
+	pvcSpec := syncRes.pvcSpec
+	datavolume := syncRes.dvMutated
+	transferName := getTransferName(datavolume)
+
+	pvcPopulated := pvcIsPopulated(pvc, datavolume)
+	_, prePopulated := datavolume.Annotations[cc.AnnPrePopulated]
+
+	if pvcPopulated || prePopulated {
+		return syncRes, nil
+	}
+
+	nn := types.NamespacedName{Namespace: datavolume.Spec.Source.Snapshot.Namespace, Name: datavolume.Spec.Source.Snapshot.Name}
+	snapshot := &snapshotv1.VolumeSnapshot{}
+	if err := r.client.Get(context.TODO(), nn, snapshot); err != nil {
+		return syncRes, err
+	}
+	if snapshot.Status == nil || snapshot.Status.ReadyToUse == nil || !*snapshot.Status.ReadyToUse {
```

Comment:
I think `== false` is cleaner than `!*`

Reply:
https://staticcheck.io/docs/checks#S1002 yells that's why I went there